spec/models/solidus_subscriptions/installment_spec.rb

```diff
@@ -168,71 +168,93 @@
-          create_list(:installment_detail, 2, installment: installment, success: false)
-
-          installment.payment_failed!(create(:order))
+          subscription = create(:subscription)
+          _last_successful_installment = create(
+            :installment,
+            subscription: subscription,
+            details: [create(:installment_detail, :success, created_at: 4.days.ago)]
+          )
+          current_installment = create(:installment, subscription: subscription)
+          current_installment.payment_failed!(create(:order))
 
-          expect(installment.subscription.state).to eq('canceled')
+          expect(current_installment.subscription.state).to eq('canceled')
         end
       end
 
       context 'when the installment has not reached the maximum number of attempts' do
         it 'creates a new installment detail' do
           allow(SolidusSubscriptions.configuration).to receive_messages(
-            maximum_reprocessing_attempts: 3,
+            maximum_reprocessing_time: 3.days,
             reprocessing_interval: 2.days,
           )
-          installment = create(:installment)
-          create_list(:installment_detail, 1, installment: installment, success: false)
-
-          installment.payment_failed!(create(:order))
+          subscription = create(:subscription)
+          _last_successful_installment = create(
+            :installment,
+            subscription: subscription,
+            details: [create(:installment_detail, :success, created_at: 1.day.ago)]
+          )
+          current_installment = create(:installment, subscription: subscription)
+          current_installment.payment_failed!(create(:order))
+          current_installment.payment_failed!(create(:order))
 
-          expect(installment.details.count).to eq(2)
+          expect(current_installment.details.count).to eq(2)
         end
 
         it "advances the installment's actionable_date" do
           allow(SolidusSubscriptions.configuration).to receive_messages(
-            maximum_reprocessing_attempts: 3,
+            maximum_reprocessing_time: 3.days,
             reprocessing_interval: 2.days,
           )
-          installment = create(:installment)
-          create_list(:installment_detail, 1, installment: installment, success: false)
-
-          installment.payment_failed!(create(:order))
+          subscription = create(:subscription)
+          _last_successful_installment = create(
+            :installment,
+            subscription: subscription,
+            details: [create(:installment_detail, :success, created_at: 1.day.ago)]
+          )
+          current_installment = create(:installment, subscription: subscription)
+          current_installment.payment_failed!(create(:order))
 
-          expect(installment.actionable_date).to eq((Time.zone.now + 2.days).beginning_of_minute)
+          expect(current_installment.actionable_date).to eq((Time.zone.now + 2.days).beginning_of_minute)
```

Comment:
I would also add a test here to ensure that the subscription does not get cancelled in this case.